### ISSUE TITLE
Enable Search By Multiple Scopes to enable sorting across Scope sorts (archives)

### ISSUE THREAD

**Issue by clarkepeterf** [project-lux/lux-marklogic#318]
**Problem Description:**
Users may want to conduct a search for records across multiple scopes. However, this is not currently possible.

One example of a scenario where one might want to do this is when searching for the members of an archival set. Right now, the frontend searches and sorts Objects and Works separately, because there is no possible way to search for them at the same time. (as was implemented in https://github.com/project-lux/lux-frontend/issues/27) However, searching and sorting for them together would allow for the correct sorting order.

**Expected Behavior/Solution:**

Users can search for records across multiple scopes (e.g. Objects and Works)

**Requirements:**

Other requirements TBD?
May have to have to do one or both of the following:
1. limit search terms to those that are the same in both scopes
2. Conduct 2 separate searches for each of the scopes and OR them together

**Needed for promotion:**
If an item on the list is not needed, it should be crossed off but not removed.

- [ ] Wireframe/Mockup - Mike
- [ ] Committee discussions - Sarah
- [ ] Feasibility/Team discussion - Sarah
- [ ] Backend requirements - TBD
- [ ] Frontend requirements- TBD
- [ ] Are new regression tests required for QA - Amy
- [ ] Questions
- List of questions for discussions. Answers should be documented within the issue.

**UAT/LUX Examples:**
- See FE ticket: https://github.com/project-lux/lux-frontend/issues/458

**Dependencies/Blocks:**
- Blocking: https://github.com/project-lux/lux-frontend/issues/458

**Related Github Issues:**
N/A

**Related links:**

- Bugherd: https://www.bugherd.com/projects/284041/tasks/2548

**Wireframe/Mockup:**
Place wireframe/mockup for the proposed solution at end of ticket.

**Comment by roamye** [comment]
From ITS Team Meeting 9/13: 
- Sorting is critical for archival findings and the requirement for this issue is to sort by archival sort numbers. (merging objects/works in this context) 
- set tabs would be easy to support with this 
- OR could be a search scope for archival sets and archival materials. 

This issue is a research ticket and will need to be prioritized.

**Comment by clarkepeterf** [comment]
For archive sorting, we currently have two indexes for each scope - `itemArchiveSortId` and `workArchiveSortId`. 

If we search for Objects we can sort by `itemArchiveSortId`, and if we search for Works we can use `workArchiveSortId`.  CTS only allows sorting by one index at a time.

I see two approaches to multiple-scope sorting:

1. Continue to use just CTS. Create a combined Object/Work index `archiveSortId`. CTS can use this single index for sorting.

2. We could use Optic to combine data from both of these indexes into a single column and sort the Optic rows by this combined column.

My hypothesis is that Optic will have slower performance than CTS, but may give us more flexibility without creating new indexes. I can test both of these out.

cc: @brent-hartwig

**Comment by clarkepeterf** [comment]
@kkdavis14 @azaroth42 can you think of any other scenarios where we may want to sort entities from multiple scopes? - Other than Objects and Works (Archival Sets) in an archive hierarchy?

**Comment by kkdavis14** [comment]
Will Bookbag have this issue?

**Comment by brent-hartwig** [comment]
@clarkepeterf, cts.search supports multiple a "sequence of cts.order specifications" and that support extends into LUX's SortCriteria and search endpoint.  That endpoint's `sort` parameter accepts a comma-delimited list of specifications: `sortKey1[:dir],sortKey2[:dir]`

**Comment by clarkepeterf** [comment]
@brent-hartwig I don't think this will work, as sortkey1 will only sort the Objects, then sortKey2 will only sort the Works. They are treated as separate sorts, so I think one scope would end up at the bottom of the results. We want both scopes mixed together. I can test that out too though

**Comment by brent-hartwig** [comment]
> My hypothesis is that Optic will have slower performance than CTS, but may give us more flexibility without creating new indexes. I can test both of these out.

Probably.  Would be interesting to confirm.  Multiple ways come to mind but in the end, you'll need a single column that has the value to sort on.  Given those values will come from two lexicons, you'll need to coalesce the two.  In the following (incomplete) example, the URIs from two plans are consolidated into a single column.

```
.bindAs(
  coalesced_uri,
  op.sem.coalesce(op.col(plan_a_uri), op.col(plan_b_uri))
)
```

**Comment by clarkepeterf** [comment]
@kkdavis14

> Will Bookbag have this issue? That's the only other thing that comes to mind on this topic.

Yes, I imagine we will want to search for all Agents, Objects, Works, Places, Events, Concepts, etc. combined that are part of this Set (collection) at the very least.

Any others to think of?

That could have a bigger impact on Optic performance as we'd have to join a bunch of different indexes into a single column.   But again, I will validate with some testing.

**Comment by brent-hartwig** [comment]
> Will Bookbag have this issue?

@kkdavis14, good question.  Boils down to what you want to sort by.  We have one that spans all content types: `anySortName` is configured to /json/identified_by[./classified_as/equivalent/id='http://vocab.getty.edu/aat/300451544']/content and we have language-specific ones.

**Comment by clarkepeterf** [comment]
Confirm 2 options are viable:
- CTS with a new, shared index
- Optic with existing indexes

Ruled out another option:
- CTS with multiple cts.order specifications

**Comment by clarkepeterf** [comment]
Findings look good for Optic, which would be my preference since it is more flexible (can use existing indexes instead of creating new ones for any combined-scope sorting):

- 8907 members - Largest archive -  [Yale University Library name/title card catalog](https://lux.collections.yale.edu/data/set/77f97611-f1fc-480e-b033-81f0e5ca96a0) (contains only Objects):
  - Cold start = 340ms
  - Warm start = 110ms 
- 5917 members - [Series I: Negatives and contact sheets](https://lux.collections.yale.edu/data/set/aedfe4bb-bd97-4375-adfe-3f2758d43a9b) (Contains only objects)
  - Cold start = 213ms
  - Warm start = 90ms
- 5781 members - [Series I: Alphabetical List of Individuals](https://lux.collections.yale.edu/data/set/d3a5e5d4-f9e7-4ed8-90af-0444e5fb1373) (5771 objects, 10 works)
  - Cold start = 436ms
  - Warm start = 90ms
- 4343 members - [Personal Correspondence](https://lux.collections.yale.edu/data/set/81d6b53d-fa19-448d-9bc1-ecb389295174) (2028 objects, 2315 works)
  - Cold start = 355ms
  - Warm start = 75ms

Code used in QC is here:
```
import op from '/MarkLogic/optic'

const scope1Query = cts.andQuery([
    cts.jsonPropertyValueQuery('dataType', ["LinguisticObject", "Set", "VisualItem"], ['exact']),
    cts.fieldValueQuery(
      ["workPartOfId"],
      ["https://lux.collections.yale.edu/data/set/81d6b53d-fa19-448d-9bc1-ecb389295174"],
      ["exact"],
      1
    )
  ])

const scope1Field = "workArchiveSortId"

const scope2Query = cts.andQuery([
    cts.jsonPropertyValueQuery('dataType', ["DigitalObject", "HumanMadeObject"], ['exact']),
    cts.fieldValueQuery(
      ["itemMemberOfId"],
      ["https://lux.collections.yale.edu/data/set/81d6b53d-fa19-448d-9bc1-ecb389295174"],
      ["exact"],
      1
    )
  ])

const scope2Field = "itemArchiveSortId"

function getXPathsFromField(fieldName, databaseName){
  const admin = require('/MarkLogic/admin.xqy');
  return admin.databaseGetFieldPaths(
    admin.getConfiguration(),
    xdmp.database(databaseName),
    fieldName
  )
  .toArray()
  .map((xmlObj) => {
    const xmlString = xmlObj.toString()
    const beginString = "<path>"
    const begin = xmlString.indexOf(beginString) + beginString.length
    let substring = xmlString.slice(begin)
    const endString = "</path>"
    const end = substring.indexOf(endString)
    substring = substring.slice(0, end)
    return substring
  }) 
}

const scope1DocPlan = op.fromSearch(scope1Query)
.joinDocUri('uri', op.fragmentIdCol('fragmentId'))

const scope1FieldPlan = op.fromLexicons({
  uri: cts.uriReference(),
  sortByMe: cts.fieldReference(scope1Field),
});

const scope1DocAndFieldPlan = scope1DocPlan.joinInner(scope1FieldPlan, op.on('uri', 'uri'))

const scope2DocPlan = op.fromSearch(scope2Query)
.joinDocUri('uri', op.fragmentIdCol('fragmentId'))

const scope2FieldPlan = op.fromLexicons({
  uri: cts.uriReference(),
  sortByMe: cts.fieldReference(scope2Field),
});

const scope2DocAndFieldPlan = scope2DocPlan.joinInner(scope2FieldPlan, op.on('uri', 'uri'))

const unionPlan = scope1DocAndFieldPlan.union(scope2DocAndFieldPlan)
unionPlan
  .orderBy(
    op.asc('sortByMe')
  )
  .limit(20)
  .result().toArray().map((row) => ({...row, name: cts.doc(row.uri).xpath(getXPathsFromField("anySortName","lux-content")[0])}))
  ```

**Comment by clarkepeterf** [comment]
@kamerynB @gigamorph @brent-hartwig 

My proposed way to expose this would be to allow searching across multiple scopes by passing in a `_multi` scope with an `OR` at the top of the JSON search criteria.

Sorting would be exposed similar to other sorting e.g. `sortCriteria:asc` - it will just be configured differently internal to the ML code to trigger the correct logic for multi-scope sorting.

Example JSON request here:
```
{
    "_scope": "multi",
    "OR": [
        {
            "_scope": "item",
            "memberOf": {
                "id": "https://lux.collections.yale.edu/data/set/e58ae36f-1ef5-41ab-a36c-5abb5d063f5e"
            }
        },
        {
            "_scope": "work",
            "partOf": {
                "id": "https://lux.collections.yale.edu/data/set/e58ae36f-1ef5-41ab-a36c-5abb5d063f5e"
            }
        }
    ]
}
```

If any concerns with this approach, let me know.

**Comment by azaroth42** [comment]
looks good to me!

**Comment by brent-hartwig** [comment]
@clarkepeterf, I like the approach.  Some questions and thoughts follow.  Apologies if some of this is addressed in previous comments.

When `_scope` is 'multi', would support be limited to a top-level OR where each item in the array sets `scope` to one of the values we support today?  If set to multi and this is not true, will an error be thrown?

I can't see a meaning for the search endpoint's `mayChangeScope` parameter when scope is 'multi'.

Is the facets endpoint to support multi-scope search criteria?  While continuing to limit facet requests to one at a time, I could see the endpoint recognizing 'multi' and selecting the subset of items in the OR array whose scope matches the endpoint's `scope` parameter (which would be required when the `_scope` property is 'multi').

Will the searchEstimate and searchWillMatch support 'multi'?

**Comment by clarkepeterf** [comment]
@brent-hartwig 
To respond to your questions/thoughts:
1. Yes, it will only support with a top-level OR and will throw an error if any non-top-level scopes are invalid
2. Indeed - I think we would ignore that parameter while processing a `multi` request
3. I could see us providing facets just for the scope which matches the facet scope - the main use case right now is just to get results, but we could add support for facets
4. I think we will need to add support for at least `searchWillMatch`, and I think it should be pretty easy to create a `searchEstimate` from the sum of the individual scopes

**Comment by brent-hartwig** [comment]
My preference would be to add support to the search, facets, searchEstimate, and searchWillMatch endpoints.  After implementing for the search endpoint, we may find support for the searchEstimate and searchWillMatch is --documentation and testing aside-- free.  The facets endpoint lift should be minimal: trim the criteria and require at least one item in the array with the facet's scope.  Priorities may not allow us to spend time on the other endpoints but let's at least make sure it's clear in https://github.com/project-lux/lux-marklogic/blob/main/docs/lux-backend-api-usage.md.

**Comment by clarkepeterf** [comment]
Functionality of the other endpoints (facets, searchEstimate, searchWillMatch) come for free with the work done to get search returning correct results 😎 - I will make sure documentation is updated with details on searching for multiple scopes

**Comment by roamye** [comment]
from team meeting [10/25](https://docs.google.com/document/d/1yISUgbiDWvyTcWroxv13pTyM7heSnBprkyAqdVfINso/edit?tab=t.0): 
- A  performance test on the week of 10/28 was not needed for this issue as it is still in the research state. 


Unsure if a performance test will be needed in the future.

**Comment by clarkepeterf** [comment]
Comparison of joining doc uri to a full list of results before sorting (sort once) vs. joining doc uri to sorted ,limited results then sorting again:

  | Sort Once | Sort Twice
-- | -- | --
  | Cold
  | 1048 | 1047
  | 960 | 947
  | 923 | 900
  | 941 | 1003
  | 932 | 941
  | 992 | 937
  | 1089 | 967
  | 998 | 980
  | 943 | 932
  | 925 | 948
  | 971 | 938
  | 932 | 930
Average | 971.1666667 | 955.8333333
  | Warm
  | 94 | 92
  | 93 | 92
  | 100 | 97
  | 102 | 96
  | 93 | 99
  | 94 | 98
  | 92 | 95
  | 92 | 99
  | 95 | 103
  | 96 | 97
Average | 95.1 | 96.8

My testing shows the difference in timing is < 2% and  showed one is slower cold vs the other is slower warm. Therefore, for now I have decided to go with sorting once.

**Comment by clarkepeterf** [comment]
@kamerynB this ticket is done and allows for a new FE ticket to be created to improve archive sorting.

The FE can now search for both items and works that are part of an archive like this:

q:
```
{
  "_scope":"multi",
  "OR":[
    {
      "_scope":"item",
      "memberOf": {
        "id":"ARCHIVE_ID"
      }
    },
    {
      "_scope":"work",
      "partOf":{
        "id": "ARCHIVE_ID"
      }
    }
  ]
}
```

and can sort using a new multi-scope sort: `archiveSortId:asc`

Other multi-scope searches are possible, and we can create other multi-scope sort criteria if that is useful.

**Comment by roamye** [comment]
@clarkepeterf - is this connected to https://github.com/project-lux/lux-frontend/issues/80 or https://github.com/project-lux/lux-frontend/issues/396 or https://github.com/project-lux/lux-frontend/issues/367 ?

**Comment by clarkepeterf** [comment]
@roamye No, it's not connected. We will need a new frontend ticket to "Display Archive Hierarchies Using Multi-Scope Search"

**Comment by brent-hartwig** [comment]
@clarkepeterf, I heard back from Engineering.  Key takeaways:

1. Pretty much the only operations that are guaranteed not to change the row order are offset and limit.
2. And while that means we should perform all joins before paginating (or expect to sort again), the optimizer has ways around it, which your time comparisons confirm.  At least during hash joins (inclusive of scatter joins), a join table is created for one plan and the results of the other table are streamed.  This is when pagination can factor in: once enough results have been streamed, the rest of the join is aborted.  So a little bit more of the join may be streamed but not an excessive amount.

**Comment by roamye** [comment]
Closing, but wanted to note this unreviewable and will need to be reviewed through the front end: https://github.com/project-lux/lux-frontend/issues/458 once that issue is complete.
